File: python/src/azoth/eos/reference/rackett_molar_volume.py
"""``eos.rackett_molar_volume`` - the saturated liquid molar volume from the Rackett
equation.

```text
v = (R*Tc/Pc)*Z_RA**(1 + (1 - Tr)**(2/7)),  Z_RA = 0.29056 - 0.08775*omega
```

Spec: ``specs/calcs/eos/rackett_molar_volume.toml``.
"""

from __future__ import annotations

from azoth._registry_gen import spec as _spec_for
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import RackettMolarVolumeResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.reference.pr_molar_volume import MOLAR_GAS_CONSTANT

CALC_ID = "eos.rackett_molar_volume"


def rackett_molar_volume(omega: float, Tc: Q, Pc: Q, T: Q) -> RackettMolarVolumeResult:
    """The saturated liquid molar volume of a pure component, from the Rackett equation.

    Args:
        omega: the Pitzer acentric factor. Dimensionless.
        Tc: critical temperature.
        Pc: critical pressure.
        T: absolute temperature, below the critical temperature.

    Returns:
        The saturated liquid molar volume, in ``m**3/mol``.

    Raises:
        OutOfRangeError: if ``Tc``, ``Pc`` or ``T`` is not positive.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = rackett_molar_volume(0.152, q(369.83, "K"), q(4_248_000.0, "Pa"), q(298.15, "K"))
        >>> round(r.v.magnitude, 18)
        8.991467403942754e-05
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    values = {
        "omega": omega,
        "Tc": input_to_si(spec, "Tc", Tc),
        "Pc": input_to_si(spec, "Pc", Pc),
        "T": input_to_si(spec, "T", T),
    }

    apply_checks(checks.on_input, values.get, warnings)

    z_ra = 0.29056 - 0.08775 * omega
    tr = values["T"] / values["Tc"]
    exponent = 1.0 + (1.0 - tr) ** (2.0 / 7.0)
    v = MOLAR_GAS_CONSTANT * values["Tc"] / values["Pc"] * z_ra**exponent

    apply_checks(checks.derived, lambda name: v if name == "v" else None, warnings)

    return RackettMolarVolumeResult(v=from_si(v, "m**3/mol"), warnings=tuple(warnings))

File: crates/azoth-eos/src/rackett_molar_volume.rs
//! `eos.rackett_molar_volume` - the saturated liquid molar volume from the Rackett
//! equation.
//!
//! ```text
//! v = (R*Tc/Pc)*Z_RA**(1 + (1 - Tr)**(2/7)),  Z_RA = 0.29056 - 0.08775*omega
//! ```
//!
//! Spec: `specs/calcs/eos/rackett_molar_volume.toml`, which records the Yamada-Gunn
//! `Z_RA` and why the exact gas constant is used.

use azoth_core::units::{Pressure, ThermodynamicTemperature, cubic_meters_per_mole};
use azoth_core::{Result, apply_checks};

use crate::pr_molar_volume::MOLAR_GAS_CONSTANT;
use crate::results::RackettMolarVolumeResult;
use crate::spec_gen;

/// The saturated liquid molar volume of a pure component, from the Spencer-Danner
/// Rackett equation.
///
/// `omega`, `Tc` and `Pc` are the caller's.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `Tc`, `Pc` or `T` is not positive.
///
/// # Example
/// ```
/// use azoth_core::units::{kelvins, pascals};
/// use azoth_eos::rackett_molar_volume;
///
/// let r = rackett_molar_volume(0.152, kelvins(369.83), pascals(4_248_000.0), kelvins(298.15))?;
/// assert!((r.v.value - 8.991467403942754e-05).abs() < 1e-18);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
#[allow(non_snake_case)] // `Tc`, `Pc` and `T` are the symbols in the published equation
pub fn rackett_molar_volume(
    omega: f64,
    Tc: ThermodynamicTemperature,
    Pc: Pressure,
    T: ThermodynamicTemperature,
) -> Result<RackettMolarVolumeResult> {
    let spec = &spec_gen::RACKETT_MOLAR_VOLUME_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "omega" => Some(omega),
            "Tc" => Some(Tc.value),
            "Pc" => Some(Pc.value),
            "T" => Some(T.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let z_ra = 0.29056 - 0.08775 * omega;
    let tr = T.value / Tc.value;
    let exponent = 1.0 + (1.0 - tr).powf(2.0 / 7.0);
    let v = MOLAR_GAS_CONSTANT * Tc.value / Pc.value * z_ra.powf(exponent);

    apply_checks(
        spec.derived_checks(),
        |name| (name == "v").then_some(v),
        &mut warnings,
    )?;

    Ok(RackettMolarVolumeResult {
        v: cubic_meters_per_mole(v),
        warnings,
    })
}
